src/server/common/simple_wml.cpp

```diff
@@ -848,53 +845,46 @@ void node::apply_diff(const node& diff)
 
 	const node* deletes = diff.child("delete");
 	if(deletes != nullptr) {
-		for(attribute_list::const_iterator i = deletes->attr_.begin(); i != deletes->attr_.end(); ++i) {
-			std::pair<attribute_list::iterator,
-	                  attribute_list::iterator> range = std::equal_range(attr_.begin(), attr_.end(), i->key, string_span_pair_comparer());
+		for(const node::attribute& a : deletes->attr_) {
+			auto range = std::equal_range(attr_.begin(), attr_.end(), a.key, string_span_pair_comparer());
```

Comment:
What is the point of using auto here?

Reply:
It was meant to be part of 45ff4786624f5172706da8377fa4ff1ddc5744b3, but I missed it. Do you want me to revert this one to pair?

Reply:
Actually, I can just unpack the pair here and wherever else `equal_range` is used, but I'd like to do that separately.
```cpp
auto [begin, end] = std::equal_range(...)
```

Reply:
I want less auto not more. Replacing a type with auto makes the code strictly less readable. I also did not understand in what way 45ff478 is supposed to be an improvement.

It seems like there is unfortunately no useful way around auto with structured bindings...

Reply:
I took care here to specify the loop var types when possible (barring cases with structured bindings, of course). I do think `auto` is a strict improvement for iterators though, especially in cases when the container type is hidden behind a typedef (as is the case here). `container::iterator` is just as opaque, since you still need to go hunt down what `container` is actually defined as.

Reply:
Calling type specifications clutter is a choice I guess. In a strongly typed language types are fairly important to understanding code...

If typedefs are badly named so they obfuscate the code instead of clarify it then we should rename or remove them instead of using auto because some IDE has trouble otherwise.

If you plan on using structured bindings anyway I have no idea why you want to rewrite everything twice.